src/client.zig

```diff
@@ -143,19 +157,54 @@ pub const Client = struct {
 
     fn on_request_timeout(self: *Client) void {
         const current_request = self.request_queue.peek() orelse return;
-        self.send_request(current_request.message);
+
+        log.debug("Retrying timed out request {o}.", .{current_request.message.header});
+        self.request_timeout.stop();
+        self.retry_request(current_request.message);
     }
 
-    fn send_request(self: *Client, request_message: *Message) void {
-        self.send_message_to_replicas(request_message);
+    fn send(self: *Client, message: *Message, isRetry: bool) void {
+        if (!isRetry) self.request_number_min += 1;
+        log.debug("{} send: request_number_min={}", .{ self.id, self.request_number_min });
+        assert(message.header.valid_checksum());
+        assert(message.header.request == self.request_number_min);
+        assert(message.header.client == self.id);
+        assert(message.header.cluster == self.cluster);
+        assert(!self.request_timeout.ticking);
+
+        self.send_message_to_replicas(message);
         self.request_timeout.start();
     }
 
+    fn send_request(self: *Client, message: *Message) void {
+        self.send(message, false);
+    }
+
+    fn retry_request(self: *Client, message: *Message) void {
+        self.send(message, true);
+    }
+
     fn on_reply(self: *Client, reply: *Message) void {
-        const done = self.request_queue.pop().?;
-        done.callback(done.user_data, done.operation, reply.body());
-        self.message_bus.unref(done.message);
+        assert(reply.header.valid_checksum());
+        assert(reply.header.valid_checksum_body(reply.body()));
+
+        const queued_request = self.request_queue.peek().?;
+        if (reply.header.client != self.id or reply.header.cluster != self.cluster) {
+            log.debug("{} on_reply: Dropping unsolicited message.", .{self.id});
+            return;
+        }
```

Comment:
We assert that we have an inflight request (`queued_request`) on line 191 with the `.?`, however, if a message arrives that is unsolicited and the right thing is to ignore it, then we could fail this `.?` assertion if we have no request inflight. I'll fix this up now on my client table branch so that it's not a blocker.